language/move-prover/bytecode-to-boogie/src/spec_translator.rs

```diff
@@ -0,0 +1,317 @@
+                //   which can be achieved with a little more effort.
+                return format!("arg{}", self.get_param_index(name));
+            }
+            StorageLocation::Old(loc) => format!("old({})", self.translate_location_as_value(loc)),
+            StorageLocation::Ret => {
+                // TODO: the parser currently only handles `return` for single value. Need to
+                //   generalize for multiple ret0, ret1, ...
+                "ret0".to_string()
+            }
+            other => {
+                if self.is_address_location(other) {
+                    // Interpret as address value.
+                    format!("Address({})", self.translate_location_as_address(other))
+                } else {
+                    // Interpret as reference which we dref.
+                    // TODO: we may want to obtimize some cases here to avoid constructing
+                    //   an intermediate Reference.
+                    format!(
+                        "Dereference(m, {})",
+                        self.translate_location_as_reference(other)
+                    )
+                }
+            }
+        }
+    }
+
+    /// Translate a location as Reference.
+    fn translate_location_as_reference(&mut self, loc: &StorageLocation) -> String {
+        match loc {
+            StorageLocation::Formal(name) => {
+                format!("GetLocalReference(m_size, {})", self.get_param_index(name))
+            }
+            StorageLocation::GlobalResource { type_, address } => format!(
+                "GetResourceReference(gs, {}, {})",
+                self.translate_location_as_address(address),
+                self.translate_resource_name(type_)
+            ),
+            StorageLocation::AccessPath { base, fields } => {
+                let mut res = self.translate_location_as_reference(base);
+                for f in fields {
+                    res.push_str(&format!(
+                        "SelectField({}, {})",
+                        res,
+                        self.translate_field(f)
+                    ));
+                }
+                res
+            }
+            _ => self.error(
+                &format!("cannot translate as reference: {:?}", loc),
+                "<ref>".to_string(),
+            ),
+        }
+    }
+
+    /// Translate a location as Address.
+    fn translate_location_as_address(&mut self, addr: &StorageLocation) -> String {
+        match addr {
+            StorageLocation::TxnSenderAddress => "TxnSenderAddress()".into(),
+            StorageLocation::Address(addr) => self.translate_account_address(addr),
+            _ => self.error(
+                &format!("cannot translate as address: {:?}", addr),
+                "<addr>".to_string(),
+            ),
+        }
+    }
+
+    /// Checks whether this is a translatable address.
+    fn is_address_location(&self, addr: &StorageLocation) -> bool {
```

Comment:
Might make sense for this to be a method on `StorageLocation`?

Reply:
Perhaps, but I suggest to wait with this until the AST/translation is stable. An alternative would actually to use a different AST type, because currently, we have folded multiple logical different concepts into StorageLocation.